c++/src/kj/compat/http.c++

```diff
@@ -5796,13 +5796,15 @@ private:
   uint concurrentRequests = 0;
   kj::Function<void(uint runningCount, uint pendingCount)> countChangedCallback;
 
-  std::queue<kj::Own<kj::PromiseFulfiller<ConnectionCounter>>> pendingRequests;
+  std::queue<kj::Own<kj::PromiseFulfiller<kj::Own<ConnectionCounter>>>> pendingRequests;
   // TODO(someday): want maximum cap on queue size?
 
-  struct ConnectionCounter final {
+  struct ConnectionCounter final: public kj::Refcounted {
```

Comment:
This commit seems to add a lot of additional heap allocation which is kind of sad.

TBH I would be just fine with not counting `errorBody` here. I don't think this is a case where being "correct for general use" is important because, to be honest, `ConcurrencyLimitingHttpClient` was build for our very specific use case in the first place, and probably shouldn't even be in KJ. I'd be OK with moving this entire class into the Workers codebase...

Reply:
Indeed, it is pretty gross. If we can move `ConcurrencyLimitingHttpClient` then definite +1 to only attaching the counter to the connection stream. I'll do that here then maybe next week I'll open a separate set of PRs that moves it.